includes/class-wc-regenerate-images.php

```diff
@@ -0,0 +1,195 @@
+<?php
+/**
+ * Regenerate Images Functionality
+ *
+ * All functionality pertaining to regenerating product images in realtime.
+ *
+ * @category Images
+ * @package WooCommerce/Classes
+ * @author Automattic
+ * @version 3.3.0
+ * @since 3.3.0
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Regenerate Images Class
+ */
+class WC_Regenerate_Images {
+
+	/**
+	 * Background process to regenerate all images
+	 *
+	 * @var WC_Regenerate_Images_Request
+	 */
+	protected static $background_process;
+
+	/**
+	 * Init function
+	 */
+	public static function init() {
+		include_once( WC_ABSPATH . 'includes/class-wc-regenerate-images-request.php' );
+		self::$background_process = new WC_Regenerate_Images_Request();
+
+		// Action to handle on-the-fly image resizing.
+		add_action( 'wp_get_attachment_image_src', array( __CLASS__, 'maybe_resize_image' ), 10, 4 );
+		// Action to handle image generation when settings change.
+		add_action( 'update_option', array( __CLASS__, 'maybe_regenerate_images_background' ), 10, 3 );
```

Comment:
We could target specific options only rather than all of them with this hook: https://core.trac.wordpress.org/browser/tags/4.8/src/wp-includes/option.php#L374